grpc/stream.go

```diff
@@ -0,0 +1,312 @@
+package grpc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-grpc/lib/netext/grpcext"
+	"github.com/mstoykov/k6-taskqueue-lib/taskqueue"
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+type message struct {
+	msg []byte
+}
+
+type stream struct {
+	vu     modules.VU
+	client *Client
+
+	logger logrus.FieldLogger
+
+	methodDescriptor protoreflect.MethodDescriptor
+
+	method string
+	stream *grpcext.Stream
+
+	tagsAndMeta    *metrics.TagsAndMeta
+	tq             *taskqueue.TaskQueue
+	builtinMetrics *metrics.BuiltinMetrics
+	obj            *goja.Object // the object that is given to js to interact with the stream
+
+	done chan struct{}
+
+	writeQueueCh chan message
+
+	eventListeners *eventListeners
+}
+
+// defineStream defines the goja.Object that is given to js to interact with the Stream
+func defineStream(rt *goja.Runtime, s *stream) {
+	must(rt, s.obj.DefineDataProperty(
+		"on", rt.ToValue(s.on), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"write", rt.ToValue(s.write), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"end", rt.ToValue(s.end), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+}
+
+func (s *stream) beginStream() error {
+	req := &grpcext.StreamRequest{
+		Method:           s.method,
+		MethodDescriptor: s.methodDescriptor,
+	}
+
+	stream, err := s.client.conn.NewStream(s.vu.Context(), *req, metadata.MD{})
+	if err != nil {
+		return fmt.Errorf("failed to create a new stream: %w", err)
+	}
+	s.stream = stream
+
+	go s.loop()
+
+	return nil
+}
+
+func (s *stream) loop() {
+	ctx := s.vu.Context()
+	wg := new(sync.WaitGroup)
+
+	defer func() {
+		wg.Wait()
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	wg.Add(2)
+	go s.readData(wg)
+	go s.writeData(wg)
+
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case <-ctxDone:
+			// VU is shutting down during an interrupt
+			// stream events will not be forwarded to the VU
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+			ctxDone = nil // this is to block this branch and get through s.done
```

Comment:
```suggestion
			return
```
I'm not getting why do we need it instead to just return it.